Merge updates and installed pages

Feichtmeier opened the pull request:
Except design acknowledgement and design improvements, the following things need to be sorted out before this can go live:

- [x] add deb page
- [x] make snap updating more reliable on page changes

Snaps
![grafik](https://user-images.githubusercontent.com/15329494/216983163-11d0f2f8-6bc8-42c7-b6a7-83e8d2c259e0.png)
![grafik](https://user-images.githubusercontent.com/15329494/216983191-de39fde8-08ce-4990-bef4-ae0311b35185.png)

Packages
![grafik](https://user-images.githubusercontent.com/15329494/216989209-a742cef5-65f5-48b3-86ad-45dd25241042.png)
![grafik](https://user-images.githubusercontent.com/15329494/216989883-2ee39370-4588-4bdb-a3b9-7dfb0e64d063.png)
![grafik](https://user-images.githubusercontent.com/15329494/216983284-c33d64e5-ddc8-4c52-bb70-299d9fae48a6.png)

(note: navigation is not consistent in this screenshots because I did not remove installed/updates items yet for less merge conflicts)


Fixes #960

Feichtmeier commented:
@d-loose 

to test this you need to

```dart
PageItem(
        titleBuilder: CollectionPage.createTitle,
        builder: (context) => CollectionPage.create(context),
        iconBuilder: (c, s) => CollectionPage.createIcon(
          context: c,
          selected: s,
          badgeCount: badgeCount,
          processing: processing,
          updateCount: updateAmount,
          updateProcessing: updatesProcessing,
        ),
      ),
      ```

d-loose commented on lib/app/collection/collection_page.dart:
```diff
@@ -0,0 +1,359 @@
+        ? installedSnaps.entries
+        : installedSnaps.entries.where(
+            (element) => element.key.name.contains(searchQuery),
+          );
```
this logic should probably be in the collection model

Feichtmeier replied:
Thanks! Done now
![grafik](https://user-images.githubusercontent.com/15329494/217022426-0a1a707e-ddc8-4a03-81f6-0cd20ee5912b.png)

d-loose commented:
`simple_snap_*.dart` have a lot of overlap with some existing code - could we somehow simplify what we already have and reuse it here?

Feichtmeier commented:
> `simple_snap_*.dart` have a lot of overlap with some existing code - could we somehow simplify what we already have and reuse it here?

Absolutely.  I just don't want to change too much of the existing pages for now

Feichtmeier commented:
The bigger model and controls have all the snap channel stuff which we don't need here which takes computation time in a big list. But generally I duplicated a lot of code here which needs to be trimmed

d-loose commented:
> Absolutely. I just don't want to change too much of the existing pages for now

How about we continue this in a separate branch where we remove the old pages and replace them by the new one? Adding new pages that are still under construction and not visible in the UI can add a lot of clutter that we'd need to clean up later.

Feichtmeier commented:
> 

Not sure if I understand correctly. You want to merge this first or work in two PRs?

d-loose commented:
I wouldn't want to merge these 'invisible' changes, but rather work on them until they're ready (either in this PR branch or in a separate feature branch if it becomes too much). Keeping the current Installed/UpdatesPages side by side with the new CollectionPage seems to create too much clutter imo.

Feichtmeier commented:
> I wouldn't want to merge these 'invisible' changes, but rather work on them until they're ready (either in this PR branch or in a separate feature branch if it becomes too much). Keeping the current Installed/UpdatesPages side by side with the new CollectionPage seems to create too much clutter imo.

sure! then I remove them here, all features are in. OK?

d-loose commented on lib/app/collection/collection_model.dart:
```diff
@@ -183,7 +183,7 @@ class CollectionModel extends SafeChangeNotifier {
       return [];
     } else {
-      if (searchQuery?.isEmpty == true) {
+      if (searchQuery == null || searchQuery?.isEmpty == true) {
```
```dart
if (searchQuery?.isEmpty ?? true)
```

d-loose commented:
> sure! then I remove them here, all features are in. OK?

That would be great - even better if you could find a way to simplify and re-use the snap model :)

Feichtmeier commented:
> > sure! then I remove them here, all features are in. OK?
> 
> That would be great - even better if you could find a way to simplify and re-use the snap model :)

the problem with the snapmodel is, that it is really bloated. I made it when I was relatively new to this API and half of the code in there is very bad and now we can do this better by either adding extensions on snapd.dart snaps or with the snap_utils.dart for all the channel stuff which is really complicated, thanks to an API only made with strings :D 

However I agree that this should be somehow synced. Can we do this in another PR? I removed all the unused code now and moved the bits used to collections dir

d-loose commented:
> However I agree that this should be somehow synced. Can we do this in another PR? I removed all the unused code now and moved the bits used to collections dir

Yes, de-bloating sounds great!

Feichtmeier commented:
> > However I agree that this should be somehow synced. Can we do this in another PR? I removed all the unused code now and moved the bits used to collections dir
> 
> Yes, de-bloating sounds great!

Cool!

All tests pass locally btw, but the CI is down today for flutter it seems
![grafik](https://user-images.githubusercontent.com/15329494/217029215-3ba084b4-9b85-4f5e-bac1-6d7cbb0f3438.png)

d-loose commented:
Yes, test failures are related to server outages.

Feichtmeier commented:
Whoopsie I have forgotten the remove button for the debs :P 

in now!
![grafik](https://user-images.githubusercontent.com/15329494/217035756-633415e0-b9b2-4cc7-8c19-0c172b35f118.png)

Plus debs are now selected first if available, as before in Updates. Was a request by @madsrh at that time

d-loose commented on lib/app/collection/collection_model.dart:
```diff
@@ -0,0 +1,239 @@
+  }
+
+  final Set<PackageKitFilter> _packageKitFilters = {
+    PackageKitFilter.installed,
```
user shouldn't be able to remove `PackageKitFilter.installed` here

Feichtmeier replied:
You can't, it's disabled in the pop-up

Feichtmeier replied:
woops I had some funny flag that was not enabled :) now it is! Thanks!

Feichtmeier commented:
Nice mock-ups!

Yes we can do this now much easier. If this will be merged please create a dedicated issue

Feichtmeier commented:
@madsrh so now you would rather display the snap page first?

d-loose reviewed (approved):
Thanks a lot, LGTM now :+1:

Feichtmeier commented:
Thank you for the thorough review :+1:
